src/net/p2p.rs

```diff
@@ -3028,11 +3050,355 @@ impl PeerNetwork {
+        }
+    }
+
+    /// Ask the remote peer for its mempool, connecting to it in the process if need be.
+    /// sets self.mempool_sync_event_id
+    /// Returns Ok(true) if we're done mempool syncing
+    /// Returns Ok(false) if there's more to do
+    fn mempool_sync_send_query(
+        &mut self,
+        url: UrlString,
+        addr: SocketAddr,
+        mempool: &MemPoolDB,
+        chainstate: &mut StacksChainState,
+    ) -> Result<bool, net_error> {
+        let sync_data = mempool.make_mempool_sync_data()?;
+        let request = HttpRequestType::MemPoolQuery(
+            HttpRequestMetadata::from_host(PeerHost::from_socketaddr(&addr)),
+            sync_data,
+        );
+
+        let event_id =
+            self.connect_or_send_http_request(url, addr, request, mempool, chainstate)?;
+        self.mempool_sync_event_id = Some(event_id);
+        return Ok(false);
+    }
+
+    /// Receive the mempool sync response.
+    /// Return Ok(true) if we're done with the mempool sync.
+    /// Return Ok(false) if we have more work to do.
+    fn mempool_sync_recv_response(&mut self, event_id: usize) -> Result<bool, net_error> {
+        PeerNetwork::with_http(self, |network, http| {
+            match http.get_conversation(event_id) {
+                None => {
+                    if http.is_connecting(event_id) {
+                        debug!("Mempool sync event {} is not connected yet", event_id,);
+                        return Ok(false);
+                    } else {
+                        // conversation died
+                        debug!("Mempool sync peer hung up");
+                        network.mempool_sync_reset();
+                        return Ok(true);
+                    }
+                }
+                Some(ref mut convo) => {
+                    match convo.try_get_response() {
+                        None => {
+                            // still waiting
+                            debug!(
+                                "Mempool sync event {} still waiting for a response",
+                                event_id
+                            );
+                            return Ok(false);
+                        }
+                        Some(http_response) => match http_response {
+                            HttpResponseType::MemPoolTxs(_, txs) => {
+                                network.mempool_sync_transactions = Some(txs);
+                                return Ok(true);
+                            }
+                            _ => {
+                                warn!("Mempool sync request received {:?}", &http_response);
+                                network.mempool_sync_reset();
+                                return Ok(true);
+                            }
+                        },
+                    }
+                }
+            }
+        })
+    }
+
+    /// Do a mempool sync
+    /// Return true if we're done and can advance to the next state.
```

Comment:
Can you add more description to the rustdoc about the state transitions here? From reading the code, it seems like `do_mempool_sync` will allow the P2P state to advance after each of the following steps:

1. Picking a peer to sync with
2. Getting the data url for that peer
3. Sending the mempool query to the peer
4. Receiving data from the peer

And then once these steps have completed, it doesn't advance the P2P state? Is there a specific reason not to advance the P2P state at the end (or, to just directly invoke the next step of the mempool sync state?)

Reply:
Your understanding of `do_mempool_sync()` is correct.  Also, the network state machine's state is advanced in `do_network_work()`, where all such writes to `self.work_state` take place (and is the only place I allow them to happen).

I'm considering refactoring this so that the mempool sync happens "in parallel" to the network state machine, much like how the neighbor walk does.  This is because I don't want to slow down inv and block syncs any more than need be.

Reply:
I went ahead and refactored this so that the mempool state machine runs in parallel to the main state machine, like how neighbor walks do